jailer.go

```diff
@@ -0,0 +1,376 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package firecracker
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+)
+
+// SecCompLevelValue represents a secure computing level type.
+type SecCompLevelValue int
+
+// secure computing levels
+const (
+	// SecCompLevelDisable is the default value.
+	SecCompLevelDisable = SecCompLevelValue(0)
+	// SecCompLevelBasic prohibits syscalls not whitelisted by Firecracker.
+	SecCompLevelBasic = SecCompLevelValue(1)
+	// SecCompLevelAdvanced adds further checks on some of the parameters of the
+	// allowed syscalls.
+	SecCompLevelAdvanced = SecCompLevelValue(2)
+)
+
+// JailerConfig is jailer specific configuration needed to execute the jailer.
+type JailerConfig struct {
+	GID           *int
+	UID           *int
+	ID            *string
+	NumaNode      *int
+	ExecFile      *string
+	ChrootBaseDir string
+	NetNS         string
+	Daemonize     bool
+	SecCompLevel  SecCompLevelValue
+}
+
+// JailerCommandBuilder will build a jailer command. This can be used to
+// specify that a jailed firecracker executable wants to be run on the Machine.
+type JailerCommandBuilder struct {
+	id       string
+	uid      int
+	gid      int
+	execFile string
+	node     int
+
+	// optional params
+	chrootBaseDir string
+	netNS         string
+	daemonize     bool
+	secCompLevel  SecCompLevelValue
+
+	stdin  io.Reader
+	stdout io.Writer
+	stderr io.Writer
+}
+
+// Args returns the specified set of args to be used
+// in command construction.
+func (b JailerCommandBuilder) Args() []string {
+	args := []string{}
+	args = append(args, b.ID()...)
+	args = append(args, b.UID()...)
+	args = append(args, b.GID()...)
+	args = append(args, b.ExecFile()...)
+	args = append(args, b.NumaNode()...)
```

Comment:
What if user forgets to provide something? Will this make incorrect command line?

Reply:
Validation handles this. Only way this would fail is if the user disables or removes the validation.